src/persist-client/src/impl/compact.rs

```diff
@@ -0,0 +1,294 @@
+            return;
+        }
+
+        let cfg = self.cfg.clone();
+        let blob = Arc::clone(&self.blob);
+        let metrics = Arc::clone(&self.metrics);
+        let mut machine = machine.clone();
+
+        // Spawn compaction in a background task, so the write that triggered it
+        // isn't blocked on it.
+        let compact_span =
+            debug_span!(parent: None, "compact::apply", shard_id=%machine.shard_id());
+        compact_span.follows_from(&Span::current());
+
+        let _ = mz_ore::task::spawn(
+            || "persist::compact::apply",
+            async move {
+                metrics.compaction.started.inc();
+                let start = Instant::now();
+                let res =
+                    Self::compact::<T, D>(cfg, Handle::current(), blob, Arc::clone(&metrics), req)
+                        .await;
+                metrics
+                    .compaction
+                    .seconds
+                    .inc_by(start.elapsed().as_secs_f64());
+
+                let res = match res {
+                    Ok(res) => res,
+                    Err(err) => {
+                        metrics.compaction.failed.inc();
+                        warn!("compaction for {} failed: {:#}", machine.shard_id(), err);
+                        return;
+                    }
+                };
+                let applied = machine
+                    .merge_res(FueledMergeRes { output: res.output })
+                    .await;
+                if applied {
+                    metrics.compaction.applied.inc();
+                } else {
+                    metrics.compaction.noop.inc();
+                }
+            }
+            .instrument(compact_span),
+        );
+    }
+
+    pub async fn compact<T, D>(
+        cfg: PersistConfig,
+        handle: Handle,
+        blob: Arc<dyn Blob + Send + Sync>,
+        metrics: Arc<Metrics>,
+        req: CompactReq<T>,
+    ) -> Result<CompactRes<T>, anyhow::Error>
+    where
+        T: Timestamp + Lattice + Codec64,
+        D: Semigroup + Codec64,
+    {
+        let compact_blocking = move || {
+            let mut parts = BatchParts::new(
+                cfg.batch_builder_max_outstanding_parts,
+                Arc::clone(&metrics),
+                req.shard_id,
+                req.desc.lower().clone(),
+                Arc::clone(&blob),
+                &metrics.compaction.batch,
+            );
+
+            // TODO: Do this in a bounded amount of memory. The pattern for how
+            // to do this has already been established in old persist (see
+            // background.rs). We'll have to generalize that from 2 to N batches
+            // and also resolve the (much much harder) issue of new persist
+            // batch sorted-ness.
+            let mut updates = Vec::new();
+            for part in req.inputs.iter() {
+                for key in part.keys.iter() {
+                    handle.block_on(fetch_batch_part(
```

Comment:
it gets a lot harder once we are doing an N way merge with bounded memory. it's a good idea, but i'd like to avoid adding complexity here before we do the bounded memory bit so we don't get in our own way :)